### ISSUE TITLE
ORBITAT provides incorrect position/velocity/true anomaly

### ISSUE THREAD

**Issue by johnwhall** [KSP-KOS/KOS#1367]
It returns the current position/velocity/true anomaly instead.

```
print ship:orbit:position.
// V(-92.5506564583629, 45.2189588607289, 115.282503618859)
print orbitat(ship, time:seconds + 60):position.
// V(-92.5506564583629, 45.2189588607289, 115.282503618859)

print ship:orbit:velocity.
// OrbitableVelocity(
//  :orbit=V(-5448.34433373436, 2743.37341503479, 4748.02053736287),
//  :surface=V(-5063.84820051011, 2743.37341503479, 4490.41164308264)
)
print orbitat(ship, time:seconds + 60):velocity.
// OrbitableVelocity(
//   :orbit=V(-5448.34433373436, 2743.37341503479, 4748.02053736287),
//  :surface=V(-5063.84820051011, 2743.37341503479, 4490.41164308264)

print ship:orbit:trueanomaly.
// 226.956461984762
print orbitat(ship, time:seconds + 60):trueanomaly.
// 226.956461984762
```

(and the orbital period wasn't 60 seconds)

**Comment by GER-Space** [comment]
This explains my Problems with the launch angle for interplanetary missions.

**Comment by hvacengi** [comment]
I've been thinking about this post for the last 2 days and trying to decide how I feel about it.  My first reaction was that these are all time dependent parameters and that the orbit itself stores no information about "time" so "of course" they aren't going to work right.  If they did, storing the orbit in a variable wouldn't return expected results:
```
set o to ship:orbit.
print o:trueanomaly.
wait o:period / 2.
print o:trueanomaly.
```
The 2nd print would be identical to the first if the orbit value included it's "time", but as implemented now the trueanomaly changes as time passes.  So making a change will have potential to break a lot of existing scripts.  And I don't think it's a good idea to make some versions of `orbit` include time info, while others do not.

The developers have also discussed a number of times just what level of "prediction" to include in kOS.  `orbitat` is really intended to determine the orbital parameters of a future orbital parameters, similar to `orbit:nextpatch`.  While there are 3 existing functions (`positionat`, `velocityat`, and `orbitat`) the current intention is to specifically not provide a `trueanomalyat` function.  You can actually predict the true anomaly at a given time with the information provided in kOS currently (I have multiple scripts that rely on this fact already).  But that has us talking out of both sides of our mouth because the same could be said for `positionat`, `velocityat`, and `eta` as well (`orbitat` can technically be done, but the SoI transition is much more complicated to predict/calculate).

All other discussion aside, the values you observed are incorrect.  They are not accurate in the orbit's frame of reference in the future, nor are they correct in that orbit's frame of reference at the "current" time (unless the future orbit is the same as the current orbit).  There should be a way for users to either determine that the information is invalid, or we should bite the bullet and provide valid information somehow.  Obviously position and velocity are already handled in other functions, perhaps the best answer is to provide the prediction function for true anomaly as well for parity.

I think the conclusion I'm at is that we need some more discussion to decide how to proceed, but that it should be addressed somehow.

**Comment by GER-Space** [comment]
for my part it is no problem that the values are not updated. It was just unexpected. A warning in the documentation and clarification about the intented use of this function should be enough (for me).

**Comment by johnwhall** [comment]
Yes, if the intention was never really to provide a true anomaly at a specified time, that's fine.  It just wasn't clear to me from the documentation.  If you think it would be helpful to others, though, I think the best solution to the lack of clarity here is to provide an Orbit class as well as an OrbitAtTime class (or something similar).  That way, kOS can return an Orbit when it wants to make it clear to users that there will be no time-dependent information, and return an OrbitAtTime when it wants to provide time-dependent information.

I also found this by chance.  I was attempting to compare kOS's orbital prediction with some external Java code I wrote, which reiterates that it's not a necessity at all.  It would be a convenience though, and certainly faster in C# than in Kerboscript.

**Comment by Dunbaratu** [comment]
The base KSP API contains values in it for the orbit object that are time dependant: altitude, trueanomaly, etc.  If that's bogus and it's actually returning values that are bull about those things, then kOS"s version of that is returning it at a frozen time rather than the current time, then that's where the fix needs to go.  We should calculate the correct time-dependant values and return them instead of KSP's own values if KSP's values are a lie.

**Comment by lamont-granquist** [comment]
@hvacengi a method like `positionat(ut)` could be added to orbit objects.  then the orbit itself wouldn't need to store a time, but you could query the orbit and give it a time and it would give you the position/velocity/anomalies/etc at the supplied ut time.

i don't know if it would mess up anything else, but you could also consider extending `position` to optionally take a UT time argument with the default being time:seconds.  that would make the most sense to me because then `orbitat(ut):position` becomes `orbitat(ut):position(time:seconds)` and you have to use `orbitat(ut):position(ut)` to get the position then -- with the explanation being more or less what you just described.  the first time parameter is used to select the orbit patch, the second ut is used to select the position out of the orbit.

and i definitely went looking for orbitat(ut):position to give me this.
